### ISSUE TITLE
Characters missing in game font due to lack of fall-back options (GNA #24621)

### ISSUE THREAD

**Issue by wesnoth-bugs** [wesnoth/wesnoth#1549]
**[Original submission](https://gna.org/bugs/index.php?24621)** by gfgtdf on 2016-04-29

Since pr #639 was merged several dialog show squares instead of the correct charactes, this can be observerd in:

1) The langugage dialog

2) Descrition text of some umc campaigns in the campaings menu

3) The hotkey preferences menu (as also shown in the discussion of pr #639 in https://cloud.githubusercontent.com/assets/4020484/14193188/b48b4800-f770-11e5-9880-ba690d934a9c.png)

(Reproduced on windows)
Release: 1.13.4+dev
Priority: 5 - Normal
Severity: 4 - Important

**Comment by wesnoth-bugs** [comment]
**Modified on 2016-08-29**

**wedge009 wrote:**

> Hot-key markers have been resolved for a while now, but issue is still present in Language options. Predominately Asian languages, I believe.

**Comment by wesnoth-bugs** [comment]
**Modified on 2017-03-26**

**ratarmy added attachment:** [japanese_version_info.PNG](http://gna.org/bugs/download.php?file_id=29902)

**Comment by wesnoth-bugs** [comment]
**Modified on 2017-03-26**

**ratarmy added attachment:** [japanese_help.PNG](http://gna.org/bugs/download.php?file_id=29903)

**Comment by wesnoth-bugs** [comment]
**Modified on 2017-03-26**

**ratarmy wrote:**

> Japanese characters are not displayed correctly in most places.
> 
> Here is a screentshot of the version information dialog in Japanese.
> 
> file #29902
> 
> 
> 
> But, in-game help is fine.
> 
> file #29903

**Comment by wesnoth-bugs** [comment]
**Modified on 2017-03-26**

**wedge009 changed status:** None -> Confirmed

**wedge009 wrote:**

> Yep, I've seen that before.
> 
> 
> 
> Marking as confirmed - probably didn't have ability to do so previously. It is indeed Windows-specific - at least in standard English Windows 7 and earlier, multi-lingual font support isn't really there by default.

**Comment by wesnoth-bugs** [comment]
**Modified on 2017-04-26**

**ratarmy added attachment:** [japanese_font_config_modified.PNG](http://gna.org/bugs/download.php?file_id=29944)

**Comment by wesnoth-bugs** [comment]
**Modified on 2017-04-26**

**ratarmy wrote:**

> Why does the family_order option in data/hardwired/fonts.cfg include only Lato family?
> 
> I changed the option to "Lato,Andagii,Droid Sans Japanese,Droid Sans Fallback", then all characters were displayed fine.
> 
> file #29944
> 
> 
> 
> I found some informations for a bug that related to family_order, but there is not such issue at least in my environment.
> 
> https://github.com/wesnoth/wesnoth/pull/638/commits/c909e41acd975c26e9d55094d1908469fbb9c7ef
> 
> https://mail.gna.org/public/wesnoth-i18n/2016-02/msg00000.html

**Comment by wesnoth-bugs** [comment]
**Modified on 2017-04-26**

**gfgtdf wrote:**

> >Why does the family_order option in data/hardwired/fonts.cfg include only Lato family? 
> 
> 
> 
> iirc this is beacsue font substitution often just isn't working, see for example https://bugzilla.gnome.org/show_bug.cgi?id=162681

**Comment by CelticMinstrel** [comment]
Surprised that the comments say this is Windows-specific... in any case, was it ever fixed? Something like this should be a blocker for 1.14.

**Comment by Wedge009** [comment]
I don't recall seeing this in Linux, only Windows. It is indeed still an issue for Windows builds, at least for me.

Love how GH says you simultaneously removed and added the windows label.

**Comment by CelticMinstrel** [comment]
I removed it because it didn't seem Windows-specific, then re-added it because one of the comments said it was.

**Comment by CelticMinstrel** [comment]
It does seem to me that Droid Sans should be included in the font list though, as @fujimo-t pointed out. I don't quite understand why that pango link means we should omit fallback fonts.

**Comment by CelticMinstrel** [comment]
1. Reading through the bug report @gfgtdf linked, I see some suggestions of ways to fix or work around this issue, either using a pango.aliases file or by passing (what sounds like) a comma-separated list to PangoFontDescription.

2. From discussion with @Vultraz it seems that adding the missing fallback fonts (as @fujimo-t mentioned) creates a different Windows issue whereby whitespace is collapsed in a manner similar to HTML / XML. In my opinion, if we have to choose between shipping with this issue and shipping with the whitespace issue, the latter would be far preferable.

**Comment by Wedge009** [comment]
Is the 'white-space issue' referring to @fujimo-t's #1572?

I think I would agree. Between that one and the missing fonts, I would prefer this missing fonts one to be resolved. Of course, that's if they must be mutually exclusive (for now).

**Comment by jyrkive** [comment]
I wonder if the font list could be dynamic, i.e. contain the fallback font if, and only if, the game is set to a language that uses non-Latin characters (Japanese, Cyrillic, Greek, etc.)

**Comment by Vultraz** [comment]
No, since there are characters not available in the Lato font that need the fallback fonts.

**Comment by gfgtdf** [comment]
I just wanted to bring back my research that i did back then on GNA #21649 (Consecutive line breaks not rendered as expected with Pango/Cairo) that was 'fixed' by making wesnoth use a single font (Lato)

> I investigated this issue, and this is related to the fontloading code: the function (inside pango) that calculates the size of empty lines passes a comma seperated list of fonts to a function that can only handle single fonts.
> Here is a stacktrace:
> ```
> pangowin32-1-vs10.dll!pango_win32_font_map_load_font(_PangoFontMap * fontmap, >_PangoContext * context, const _PangoFontDescription * description)  Zeile 882	C
> pango-1-vs10.dll!pango_font_map_load_font(_PangoFontMap * fontmap, _PangoContext * context, const _PangoFontDescription * desc)  Zeile 95 + 0x18 Bytes	C
> pango-1-vs10.dll!pango_context_load_font(_PangoContext * context, const _PangoFontDescription * desc)  Zeile 319 + 0x14 Bytes	C
> pango-1-vs10.dll!pango_layout_get_empty_extents_at_index(_PangoLayout * layout, int index, _PangoRectangle * logical_rect)  Zeile 4550 + 0x10 Bytes	C
> pango-1-vs10.dll!pango_layout_line_get_empty_extents(_PangoLayoutLine * line, _PangoRectangle * logical_rect)  Zeile 4590 + 0x16 Bytes	C
> pango-1-vs10.dll!pango_layout_line_get_extents(_PangoLayoutLine * line, _PangoRectangle * ink_rect, _PangoRectangle * logical_rect)  Zeile 4819 + 0xd Bytes	C
> pango-1-vs10.dll!get_line_extents_layout_coords(_PangoLayout * layout, _PangoLayoutLine * line, int layout_width, int y_offset, int * baseline, _PangoRectangle * line_ink_layout, _PangoRectangle * line_logical_layout)  Zeile 2551 + 0x1a Bytes	C
> pango-1-vs10.dll!pango_layout_get_extents_internal(_PangoLayout * layout, _PangoRectangle * ink_rect, _PangoRectangle * logical_rect, _GSList * * line_extents)  Zeile 2665 + 0x30 Bytes	C
> pango-1-vs10.dll!pango_layout_get_extents(_PangoLayout * layout, _PangoRectangle * ink_rect, _PangoRectangle * logical_rect)  Zeile 2785 + 0x13 Bytes	C
> pango-1-vs10.dll!pango_layout_get_pixel_extents(_PangoLayout * layout, _PangoRectangle * ink_rect, _PangoRectangle * logical_rect)  Zeile 2811 + 0x11 Bytes	C
> ```

**Comment by CelticMinstrel** [comment]
@Wedge009 - From what I've heard from @Vultraz, the whitespace issue isn't #1572 but rather (as I described in the previous comment) an issue with consecutive newlines or spaces being collapsed into a single one.

**Comment by Vultraz** [comment]
Indeed, it isn't #1572.

**Comment by jyrkive** [comment]
@gfgtdf When did you do that research? We use a newer version of Pango now, especially in MSVC builds.

**Comment by gfgtdf** [comment]
iirc it was pango 1.38.1

**Comment by jyrkive** [comment]
Regarding my earlier comment:

> I wonder if the font list could be dynamic, i.e. contain the fallback font if, and only if, the game is set to a language that uses non-Latin characters (Japanese, Cyrillic, Greek, etc.)

A bit of research showed that the list is already dynamic. It's marked as translatable - therefore translators can already override it.

I don't understand @Vultraz's response to my comment. `family_order` contains **only** the Lato font in the English version.

So, my suggestion on how to improve the situation is that each translator overrides `family_order` if their language doesn't use Latin letters. Suggested value:

`Lato,DejaVu Sans,Andagii,Droid Sans Japanese,Droid Sans Fallback,Junicode`

**Comment by gfgtdf** [comment]
> situation is that each translator overrides `family_order` if their language doesn't use Latin letters. Suggested value:

As noted in the initial post, non-english letters can also appear even though your are playing the game in english

> `Lato,DejaVu Sans,Andagii,Droid Sans Japanese,Droid Sans Fallback,Junicode`

Whe doing that, please check whethe it brings GNA #21649  back.

**Comment by jyrkive** [comment]
> Whe doing that, please check whethe it brings GNA #21649 back.

I did check. Unfortunately, the change does bring the issue back. :(

I believe that my suggestion would be the best compromise we have. The current situation is that non-Latin characters are completely broken on Windows regardless of which language is used - even if the language consists of exclusively such characters. That's just unacceptable.

@fujimo-t's suggestion to change `family_order` for all languages would get non-Latin characters working for all languages, but at the cost of collapsing empty lines.

If `family_order` is only changed for languages which need it, the game would be at least playable for everyone. Some players would get "non-Latin characters broken on Windows" and others would get "empty lines disappear", but no one would get a completely broken game experience, and only those who need non-Latin characters would get the whitespace issue.

**Comment by Vultraz** [comment]
For the record, it's not just empty lines that get collapsed. The addition of the rest of the family_order list also weirdly affected whitespace calculations within a line. You'll see dialogs like the Langauge shrinik actually, IIRC)

**Comment by CelticMinstrel** [comment]
Even so, mishandling of whitespace is still far preferable to being unable to see any of the text.

**Comment by AI0867** [comment]
Could it be a 'solution' to have a separate `windows_family_order` that contains just a single font, and which is translator-hinted to only use the best available font for that language?

**Comment by Pentarctagon** [comment]
Removed the Blocker tag, since clearly this isn't blocking releases.

**Comment by Wedge009** [comment]
Testing this against current Steam release 1.14.6, switching to Japanese language in Windows, I don't have this issue any more (checked Preferences, Version dialogue, and Help documentation). Only minor issue is when using a latin-based language and looking in the Language dialogue, the Chinese and Japanese languages listed in their native characters aren't rendered (but they are when Japanese language is selected).

Can this mean this is resolved now?

**Comment by gfgtdf** [comment]
> checked Preferences, Version dialogue, and Help documentation

Iirc another dialog where this showed up was the credits screen.

**Comment by Wedge009** [comment]
Credits is fine under 1.14.6 with Japanese language selected, other than for Marathi - I expect that one probably needs an Indian font. I still don't think that's enough to keep this issue open.

**Comment by jyrkive** [comment]
The Japanese translation is using the workaround I suggested in https://github.com/wesnoth/wesnoth/issues/1549#issuecomment-323571780:

https://github.com/wesnoth/wesnoth/blob/97b1f7630cadbb58710dc492a6ae65d8fe513e84/po/wesnoth/ja.po#L4043-L4046

Whether to close this issue or not depends on whether we consider this workaround an acceptable fix.

**Comment by gfgtdf** [comment]
> I think this can be closed, but re-open if you feel otherwise

The problem is that 1) Its a regression, 2) Might also effect addon code, but more importantly 3) a reminder of the 

> iirc this is because font substitution often just isn't working, see for example https://bugzilla.gnome.org/show_bug.cgi?id=162681

part, which i still think is the real problem here, becasue otherwise we coudl just do what teh seocnd comment said and add fallback fonts.


> Whether to close this issue or not depends on whether we consider this workaround an acceptable fix.

whether its a acceptable fix depend on whether it works in paricular on windows which the gome bug suggested works not (or at least has other side effects, iirc i tested it once oen windows and adding fallback fonts caused addintional nwelined to be inderted in text)

**Comment by Wedge009** [comment]
What exactly is the regression? What is there still to be covered besides a bit of Marathi text? Are we going to add Indian fonts too?

I thought that's what we've done for the CJK texts.

I have checked Windows, as well as Linux - what more is there that's missing that still needs to be resolved?

**Comment by jyrkive** [comment]
@Wedge009 The primary problem is that empty lines get collapsed with languages which need to use the workaround, such as Chinese and Japanese. That's the reason why we aren't using the modified font list with all languages.

And a secondary problem is that this also affects us when we want to show international characters while Wesnoth itself uses a Western locale (e.g. player usernames or add-on descriptions).

**Comment by Wedge009** [comment]
I'm aware that there are issues with lines and spacing from ##1572. But I haven't seen empty lines being 'collapsed'. Is that still happening now? Because I can't see it with 1.14.7. But isn't this a separate issue anyway?

From the original description I took this issue to mean that squares were being used as place-holders for missing characters in the current fonts that we're using. Honestly, I don't remember observing Wesnoth to be completely 'square free' in Windows in the past so I don't know what the 'regression' is. Currently I see very few squares in Windows (and even fewer in Linux) so as best as I can tell this is resolved. If there are still side-issues, open new reports. If this isn't resolved, re-open it. I don't understand what the issue is here.

**Comment by gfgtdf** [comment]
> I'm aware that there are issues with lines and spacing from ##1572.

This is a differnt issue


> But I haven't seen empty lines being 'collapsed'. Is that still happening now?

Well from the preivous comments it might happen in japanse language. although im not sure whether it applies to japanaese in the first place.

> Because I can't see it with 1.14.7. But isn't this a separate issue anyway?

No the point is that this issue (characters missing in game font) is not compltely solved, we only fixed the most important cases in mainline by using different symbols. The only real fix for this issue (using backup fonts) is blocked by an upsteam bug that makes empty lines collapsed on windows if we apply it.


> Honestly, I don't remember observing Wesnoth to be completely 'square free' in Windows in the past so I don't know what the 'regression' is.


Well there are cases that show squares now that diplayed properly in 1.10, (or was it 1.12 ? not sure when this was changed), if not im mainline then surely in umc.

> I don't understand what the issue is here.

well its that there are still quares.

**Comment by gfgtdf** [comment]
Im removing the regression label for now, while its true that there might be some cases  that might be displayed worse than in 1.10, i don't think the situation is generally worse.

**Comment by Wedge009** [comment]
I checked Japanese as well and didn't see 'collapsed' lines. #1572 is still an issue, though.

I still don't see that many missing characters, but I suppose it's still happening a little. I was going to suggest adding the Upstream issue label, but you've already done that.

It sounds like we've gone beyond the original report, then.

**Comment by gfgtdf** [comment]
> I checked Japanese as well and didn't see 'collapsed' lines. #1572 is still an issue, though.

Hmm yes i'm an on linux and can not test this issue, since it appears only on windows.

**Comment by dway123** [comment]
Update: This bug still repros in the Windows+Steam release of Wesnoth version 1.14.7.

While using Wesnoth on Windows, and setting the language to English (or any other Latin-based or Cryllic) language, CJK languages (ex. Chinese and Japanese) and Marathi still fail to display correctly. This can be confirmed in the credits page, per screenshots in #5977.

Per @Wedge009's [comment](https://github.com/wesnoth/wesnoth/issues/5977#issuecomment-886003097), this still doesn't reproduces on Linux, or when setting the language to a language like Chinese or Japanese.

**Comment by Wedge009** [comment]
Thanks for testing. Did you mean 1.14.17?

**Comment by dway123** [comment]
Oops yes, I did mean 1.14.17, thanks. I've edited my previous comment to say this now.

**Comment by irydacea** [comment]
This should continue to affect 1.15.x for the foreseeable future.

There is an existing solution by @loonycyborg in the `use_fontconfig` branch that fixes both this and the issue with fallbacks causing line breaks to be collapsed, but it hasn't been merged into master because of concerns that it could break Wesnoth horribly on certain configurations* on versions of Windows prior to Windows 10 1903, which is the first version that allows selecting the UTF-8 codepage application-wide for ANSI Windows API calls. This is important because fontconfig uses those and we don't have control over the fontconfig codebase to make it use wide character (Unicode) APIs instead.

(* Namely, on any configuration where the path leading to Wesnoth's font files contains non-ASCII characters.)

**Comment by Pentarctagon** [comment]
At what point do we want to bump the minimum required version of Windows to Windows 10 then?  By the time 1.18 is released Windows 7 will be out of support for at least a couple years and Windows 8/8.1 is more or less irrelevant given its tiny market share.

**Comment by Pentarctagon** [comment]
Barring any objections, I plan to merge #6685 prior to 1.17.5.
